[blog] Qdrant Version 1.14 (#1529)
qdrant/landing_page: qdrant-landing/content/blog/qdrant-1.14.x.md

### Minor Fixes & Optimizations
![reranking](/blog/qdrant-1.14.x/gridstore.jpg)

**Optimized Memory Usage in Immutable Segments**

We revamped how the ID tracker and related metadata structures store data in memory. This can result in a notable RAM reduction for very large datasets (hundreds of millions of vectors).

This causes **much lower overhead**, where memory savings let you store more vectors on the same hardware. Also, improved scalability is a major benefit. If your workload was near the RAM limit, this might let you push further **without using additional servers**.

**IO Measurements for Serverless Deployments**

Qdrant 1.14 introduces detailed tracking of **read/write costs** (CPU, disk, etc.) per operation. This is primarily intended for serverless billing, but also helps diagnose performance hotspots in dedicated setups.

> Now you can have **full cost visibility**, and you can understand exactly which queries or updates cause the most overhead.

This also makes for easier optimization - you can tweak indexes, partitioning, or formula queries to reduce resource usage based on concrete metrics.

**Ending our Reliance on RocksDB**

> The **mutable ID tracker no longer relies on RocksDB**. This continues our journey toward minimal external dependencies. With our custom-built [**Gridstore**](/articles/gridstore-key-value-storage/), you can expect fewer random compactions and more predictable disk usage.

Reviewer: The new mutable ID tracker does **not** rely on RocksDB.

It's another kind of custom storage, which is very specialized to its purpose - the mutable ID tracker.

How it works is actually very simple: we have point mappings, and we store all changes to these mappings in a file simply by appending each change as an entry.

If we add a point, we create a mapping, thus we append a 'create point mapping for point x' entry to the file. If we delete a mapping, a 'delete point mapping for point x' is appended to the file. On load we simply walk over all change entries and reconstruct the mappings in memory.

Of course, this could grow forever if there are a lot of changes.

But, this integrates very well with our optimizers. The optimizers ensure that if the ID tracker grows large, it is picked up for optimization. In that case it is transformed into the immutable ID tracker, which benefits from other optimizations. That's why this simple mechanism in the mutable ID tracker is very effective here.

> This reduces complexity, with fewer external data engines in your stack.

Reviewer: Though it's not built on GridStore, this is definitely still true.

Your website collection can have vectors for **titles**, **paragraphs**, and **code snippet** sections of your documentation. You can create a `tag` payload field that indicates whether a point is a title, paragraph, or snippet. Then, to give more weight to titles and paragraphs, you might do something like:

```
score = score + (is_title * 0.5) + (is_paragraph * 0.25)
```

**Above is just sample logic - but here is the actual Qdrant API request:**

```bash
POST /collections/{collection_name}/points/query
{
    "prefetch": {
        "query": [0.2, 0.8, ...],   // <-- dense vector for the query
        "limit": 50
    },
    "query": {
        "formula": {
            "sum": [
                "$score",
                {
                    "mult": [
                        0.5,
                        { "key": "tag", "match": { "any": ["h1","h2","h3","h4"] } }
                    ]
                },
                {
                    "mult": [
                        0.25,
                        { "key": "tag", "match": { "any": ["p","li"] } }
                    ]
                }
            ]
        }
    }
}
```

### Idea 2: Reranking Most Recent Results

> One of the most important advancements is the ability to prioritize recency. In many scenarios, such as in news or job listings, users want to see the most recent results first. Until now, this wasn’t possible without additional work: *you had to fetch all the data and manually filter for the latest entries on their side*.

Reviewer 2: ```suggestion
One of the most important advancements is the ability to prioritize recency. In many scenarios, such as in news or job listings, users want to see the most recent results first. Until now, this wasn’t possible without additional work: *you had to fetch all the data and manually filter for the latest entries on your side*.
```

I might be wrong, but it would sound more naturally to me.

Now, the similarity score **doesn’t have to rely solely on cosine distance**. It can also take into account how recent the data is, allowing for much more dynamic and context-aware ranking. 

> With the Score-Boosting Reranker, simply add a `datetime` payload field and factor it into your formula so fresher data rises to the top.

**Example Query**:

```bash
POST /collections/{collection_name}/points/query
{
  "prefetch": { ... },
  "query": {
    "formula": {
      "sum": [
        "$score",
        {
          "gauss_decay":
            "target": { "datetime": <todays date> },
            "x": { "datetime_key": <payload key> },
            "scale": <1 week in seconds>
        }
      ]
    }
  }
}
```

### Idea 3: Factor in Geographical Proximity

> Let’s say you’re searching for a restaurant serving Currywurst. Sure, Berlin has some of the best, but you probably don’t want to spend two days traveling for a sausage covered in magical seasoning. The best match is the one that **balances the distance with a real-world geographical distance**. You want your users see relevant and conveniently located options.

Reviewer 2: ```suggestion
Let’s say you’re searching for a restaurant serving Currywurst. Sure, Berlin has some of the best, but you probably don’t want to spend two days traveling for a sausage covered in magical seasoning. The best match is the one that **balances the distance in the vector space with a real-world geographical distance**. You want your users see relevant and conveniently located options.
```

        "query": [0.2, 0.8, ...],
        "limit": 50
    },
    "query": {
        "formula": {
            "sum": [
                "$score",
                {
                    "gauss_decay": {
                        "scale": 5000,               // e.g. 5 km
                        "x": {
                            "geo_distance": {
                                "origin": { "lat": 52.504043, "lon": 13.393236 }, // Berlin
                                "to": "geo.location"
                            }
                        }
                    }
                }
            ]
        },
        "defaults": {
            "geo.location": { "lat": 48.137154, "lon": 11.576124 } // Munich
        }
    }
}
```

You can tweak parameters like `target`, `scale`, and `midpoint` to shape how quickly the score decays over distance. This is extremely useful for local search scenarios, where location is a major factor but not the only factor.

> This is a very powerful feature that allows for extensive customization. Read more about this feature in the [**Hybrid Queries Documentation**](/documentation/concepts/hybrid-queries/)

## Improved Resource Use During Segment Optimization

> ![optimizations](/blog/qdrant-1.14.x/optimizations.jpg)

Reviewer 2: ```suggestion
[![optimizations](/blog/qdrant-1.14.x/optimizations.jpg)](/blog/qdrant-1.14.x/optimizations.jpg)
```

I would like to see a bigger version of the image.

These limitations can lead to issues like latency spikes that are difficult to diagnose or mitigate. Our long-term goal is to fully eliminate RocksDB to gain complete control over Qdrant’s performance and storage behavior. **That’s also why we built GridStore**—a key-value engine designed specifically for our needs.

> The last remaining use of RocksDB is within **mutable payload indices**, and that too will be removed in a future release, fully cutting the dependency.

![reranking](/blog/qdrant-1.14.x/gridstore.png)

*Read more about how we built [**Gridstore, our custom key-value store**](/articles/gridstore-key-value-storage/).*

#### Optimized Memory Usage in Immutable Segments

We revamped how the ID tracker and related metadata structures store data in memory. This can result in a notable RAM reduction for very large datasets (hundreds of millions of vectors).

This causes **much lower overhead**, where memory savings let you store more vectors on the same hardware. Also, improved scalability is a major benefit. If your workload was near the RAM limit, this might let you push further **without using additional servers**.

#### I/O Measurements for Serverless Deployments

> Qdrant 1.14 introduces detailed tracking of **read/write costs** (CPU, disk, etc.) per operation. This is primarily intended for serverless billing, but also helpsyou diagnose performance hotspots in dedicated setups.

Reviewer 2: ```suggestion
Qdrant 1.14 introduces detailed tracking of **read/write costs** (CPU, disk, etc.) per operation. This is primarily intended for serverless billing, but also helps you diagnose performance hotspots in dedicated setups.
```

> Now you can have **full cost visibility**, and you can understand exactly which queries or updates cause the most overhead.

> This also makes for easier optimization - you can tweak indexes, partitioning, or formula queries to reduce resource usage based on concrete metrics.

Reviewer 2: The first question that comes to my mind is how to check the cost of the individual operation. Is that displayed in logs or returned in an operation response? Is that an internal feature only? If not, having a link to a detailed description here could help.

Author: waiting on this

---
title: "Qdrant 1.14 - Reranking Support & Extensive Resource Optimizations"
draft: false
short_description: "Qdrant 1.14 adds Score-Boosting Reranker for custom search ranking and improved performance optimizations."
description: "Qdrant 1.14 introduces Score-Boosting Reranker for custom search ranking, and improved resource utilization." 
preview_image: /blog/qdrant-1.14.x/social_preview.png
social_preview_image: /blog/qdrant-1.14.x/social_preview.png
date: 2025-04-14T00:00:00-08:00
author: David Myriel
featured: true
tags:
  - vector search
  - vector database
  - semantic search
  - release notes
  - performance optimization
  - reranking
  - search engine
  - database
  - AI
  - machine learning
---

[**Qdrant 1.14.0 is out!**](https://github.com/qdrant/qdrant/releases/tag/v1.14.0) Let's look at the main features for this version:

**Score-Boosting Reranker:** Blend vector similarity with custom rules and context.</br>
**Improved Resource Utilization:** CPU and disk IO optimization for faster processing.</br>

**Incremental HNSW Indexing:** Build indexes gradually as data arrives.</br>

> **Batch Search:** Optimized parallel processing for multiple queries.</br>

Reviewer: This is specific to a single request being a batch of queries. So let's call it batch queries here.

```suggestion
**Batch Search:** Optimized parallel processing for batch queries.</br>
```

**Memory Optimization:** Reduced usage for large datasets with improved ID tracking.</br>

> **RocksDB to Gridstore:** Additional reliance on our optimized key-value store.</br>

Reviewer: ```suggestion
```

See [this](https://github.com/qdrant/landing_page/pull/1529#discussion_r2047054957)

> **RocksDB to Gridstore:** Additional reliance on our custom key-value store. </br>

Reviewer: We use it because it's more 'optimized' for our use case

```suggestion
**RocksDB to Gridstore:** Additional reliance on our optimized key-value store.</br>
```

## Incremental HNSW Indexing 
![optimizations](/blog/qdrant-1.14.x/optimizations.jpg)

Rebuilding an entire [**HNSW graph**](/documentation/concepts/indexing/#vector-index) every time new data is added can be computationally expensive. With this release, Qdrant now supports incremental HNSW indexing—an approach that extends existing HNSW graphs rather than recreating them from scratch.

> This feature is designed to make indexing faster and more efficient when you’re only adding new points. It reuses the existing structure of the HNSW graph and appends the new data directly onto it. 

That means much less time spent building and more time searching. Although this initial implementation currently only support upserts, it lays the groundwork for a more dynamic and performance-friendly indexing process. Especially for collections with frequent updates to payload values or growing datasets, incremental HNSW is a big step forward.

> Note that deletes and updates will still trigger a full rebuild. Check out the [**indexing documentation**](/documentation/concepts/indexing/) to learn more. 

## Faster Batch Queries
![reranking](/blog/qdrant-1.14.x/gridstore.jpg)

> In this release, Qdrant introduces a major performance boost for [**batch search operations**](/documentation/concepts/search/#batch-search-api). Until now, the query batch API used a single thread per segment, which worked well—unless you had just one segment and a large batch of queries. In such cases, everything was processed on a single thread, significantly slowing things down. This scenario was especially common when using our [**Python client**](https://github.com/qdrant/qdrant-client), which is single-threaded by default.

Reviewer: There is a search and query endpoint, since this focuses on the new query endpoint, lets name that explicitly:

```suggestion
In this release, Qdrant introduces a major performance boost for [**batch query operations**](/documentation/concepts/search/#batch-search-api). Until now, the query batch API used a single thread per segment, which worked well—unless you had just one segment and a large batch of queries in a single request. In such cases, everything was processed on a single thread, significantly slowing things down. This scenario was especially common when using our [**Python client**](https://github.com/qdrant/qdrant-client), which is single-threaded by default.
```

The new optimization changes that. Large query batches are now split into chunks, and each chunk is processed on a separate thread. 

> > This allows Qdrant to execute queries concurrently, even when operating over a single segment. 

Reviewer: ```suggestion
> This allows Qdrant to execute queries concurrently, even when operating over a single segment with a limited number of requests.
```

You will get much faster response times for large batches of queries. If you’re working with high-volume query workloads, you should notice a significant improvement in latency. Benchmark results show just how dramatic the difference can be.

> Using our [**bfb benchmarking tool**](https://github.com/qdrant/bfb), we populated a 1-segment collection and ran a request of 240 batch queries.

Reviewer: Added 'basic test' as you might not call this very scientific. But since the difference is so dramatic its already good enough to prove there's a perf improvement. Not sure about my exact wording here though.

```suggestion
As a basic test, we populated a 1-segment collection using our [**bfb benchmarking tool**](https://github.com/qdrant/bfb), and ran a request of 240 batch queries.
```

Initially, this process only saturated a single CPU and took 11 seconds:
![parallel-before](/blog/qdrant-1.14.x/parallel-before.png)

After, it saturated 23 CPUs (search threads = CPU count - 1) and took 4.5 seconds:
![parallel-after](/blog/qdrant-1.14.x/parallel-after.png)

We ran the same test of large queries for the following configurations:

| Configuration | Segments | Shards | Indexing Threshold | Before | After | Improvement |
|--------------|----------|---------|-------------------|---------|--------|-------------|

> | Single Segment | 1 | 1 | - | 10.5s | 4.5s | 57% |
> | Multi Segment | 10 | 1 | - | 5.5s | 5.1s | 7% |
> | Single Segment, Multi Shard | 1 | 4 | - | 5.2s | 4.6s | 12% |
> | Multi Everything | 10 | 4 | 1000 | 5.1s | 4.5s | 12% |
> | High Shard, Single Segment | 1 | 16 | 1000 | 5.2s | 3.7s | 29% |
> | High Shard, Multi Segment | 10 | 16 | 1000 | 2.5s | 1.7s | 32% |

Reviewer: ```suggestion
| Single Segment | 1 | 1 | 20k | 10.5s | 4.5s | 57% |
| Multi Segment | 10 | 1 | 20k | 5.5s | 5.1s | 7% |
| Single Segment, Multi Shard | 1 | 4 | 20k | 5.2s | 4.6s | 12% |
| Multi Everything | 10 | 4 | 1k | 5.1s | 4.5s | 12% |
| High Shard, Single Segment | 1 | 16 | 1k | 5.2s | 3.7s | 29% |
| High Shard, Multi Segment | 10 | 16 | 1k | 2.5s | 1.7s | 32% |
```

This leads to **faster optimizations**, which are specially noticeable on large machines handling big data movement.
It also gives you **predictable performance**, as there are fewer sudden spikes or slowdowns during indexing and merging operations.

**Figure 2:** Indexing 400 million vectors - CPU and disk usage profiles. Previous Qdrant version on the left, new Qdrant version on the right.
![indexation-improvement](/blog/qdrant-1.14.x/indexation.png)

**Observed Results:** The new version on the right clearly shows much better CPU saturation across the full process. The improvement is especially noticeable during large-scale indexing. 

In our experiment, **we indexed 400 million 512-dimensional vectors**. The previous version of Qdrant took around 40 hours on an 8-core machine, while the new version with this change completed the task in just 28 hours.

> **Tutorial:** If you want to work with a large number of vectors, we can show you how. [**Learn how to upload and search large collections efficiently.**](/documentation/database-tutorials/large-scale-search/)

## Ending our Reliance on RocksDB
![rocksdb-gridstore](/blog/qdrant-1.14.x/rocksdb-gridstore.jpg)

> RocksDB has been removed from the **mutable ID tracker** and all **immutable payload indices**, which are both internal components of Qdrant. In practical terms, this means: less RocksDB, more customization, faster internals.

Reviewer: ```suggestion
RocksDB has been removed from the **mutable ID tracker** and all **immutable payload indices**, which are both internal components of Qdrant. In practical terms, this means: less RocksDB, more specialized storage, faster internals.
```

or

```suggestion
RocksDB has been removed from the **mutable ID tracker** and all **immutable payload indices**, which are both internal components of Qdrant. In practical terms, this means: less RocksDB, more control, faster internals.
```
